cow_amm/cow_amm_lp_and_tvl_4047078.sql

```diff
@@ -0,0 +1,151 @@
+lp_balance_delta as (
+    select
+        date(evt_block_time) as "day",
+        sum(case when "from" = 0x0000000000000000000000000000000000000000 then value else -value end) as lp_supply
+    from erc20_ethereum.evt_transfer
+    where
+        contract_address in (select address from cow_amm_pool)
+        and ("from" = 0x0000000000000000000000000000000000000000 or "to" = 0x0000000000000000000000000000000000000000)
+    group by 1
+),
+
+total_lp as (
+    select
+        day,
+        lp_supply,
+        sum(lp_supply) over (order by day) as total_lp
+    from lp_balance_delta
+),
+
+-- lp token total supply without date gaps
+lp_total_supply as (
+    select
+        day,
+        total_lp
+    from (
+        -- join full date range with potentially incomplete data. This results in many rows per day (all total supplies on or before that day)
+        -- rank() is then used to order join candidates by recency (rank = 1 is the latest lp supply)
+        select
+            date_range.day,
+            total_lp,
+            rank() over (partition by (date_range.day) order by lp.day desc) as latest
+        from date_range
+        inner join total_lp as lp
+            on date_range.day >= lp.day
+            -- performance optimisation: this assumes one week prior to start there was at least one lp supply change event
+            and lp.day >= (timestamp '{{start}}' - interval '7' day)
+    )
+    where latest = 1
+),
+
+-- Compute tvl by multiplying each day's closing price with net transfers in and out of the pool
+get_tvl as (
+    select
+        x.day,
+        sum(amount * price_close / pow(10, decimals)) as total_tvl
+    from (
+        select
+            date(evt_block_time) as "day",
+            contract_address,
+            -value as amount
+        from erc20_ethereum.evt_transfer
+        where
+            "from" in (select address from cow_amm_pool)
+
+        union all
+        select
+            date(evt_block_time) as "day",
+            contract_address,
+            value as amount
+        from erc20_ethereum.evt_transfer
+        where
+            "to" in (select address from cow_amm_pool)
+    ) as x inner join prices.usd_daily as y
+        on blockchain = 'ethereum' and x.day = y.day and x.contract_address = y.contract_address
+    group by 1
+),
+
+total_tvl_prep as (
+    select
+        day,
+        sum(total_tvl) over (order by day) as total_tvl
+    from get_tvl
+),
+
+total_tvl as (
+    select
+        day,
+        total_tvl
+    from (
+        -- join full date range with potentially incomplete data. This results in many rows per day (all total tvl on or before that day)
+        -- rank() is then used to order join candidates by recency (rank = 1 is the latest tvl)
+        select
+            date_range.day,
+            total_tvl,
+            rank() over (partition by (date_range.day) order by tvl.day desc) as latest
+        from date_range
+        inner join total_tvl_prep as tvl
```

Comment:
I created a helper query in #27 which we can use here instead of total_tvl_prep which should simplify this query quite a bit.